### PR TITLE
feat: support Vercel AI SDK v7

### LINKED ISSUE #510: Support ai v7
Heya OpenRouter team,

ai@7.0.0 (and now 7.0.1) has been released. It has some breaking changes. ESM-only, for one.
Any chance you can release an updated `@openrouter/ai-sdk-provider` release to support it?

Thanks in advance!

ref:
https://github.com/vercel/ai/releases/tag/ai%407.0.0

### PR DESCRIPTION
**Pull request by Shreyaan**
Fixes #510.

## Design choice

Chose (A) v7-only. AI SDK v7 is Node >=22 and ESM-only, and first-party v7-line providers such as `@ai-sdk/openai@4` target `@ai-sdk/provider@4` V4 interfaces. Keeping v6 support would require compatibility branching around provider/model specs, file payloads, tool factories, packaging, and usage shapes, so this is a major release.

## Summary

- Bumps `ai` to `^7`, `@ai-sdk/provider` to `^4`, and `@ai-sdk/provider-utils` to `^5`; peers on `ai@^7.0.0`.
- Migrates chat and completion language models to `LanguageModelV4`, including V4 content/file payloads, provider tools, reasoning, structured outputs, cache-control passthrough, tool calls, and stream lifecycle events.
- Migrates embedding, image, and video models to the V4 spec. Video tests now pass the required `generateAudio` call option.
- Updates OpenRouter web search to `createProviderDefinedToolFactory` / `ProviderDefinedToolFactory`.
- Publishes ESM-only output with `type: module`, Node >=22, and no CommonJS export branch.
- Preserves OpenRouter usage accounting, reasoning details, cache control/cached-token handling, tool calling, structured outputs, BYOK cost metadata, and provider options.

## Usage/token remapping

Provider usage now follows the V4 internal shape: `inputTokens.cacheRead/cacheWrite/noCache` and `outputTokens.reasoning/text`. In AI SDK v7 public results, cached input tokens are read from `usage.inputTokenDetails.cacheReadTokens` and reasoning tokens from `usage.outputTokenDetails.reasoningTokens`; OpenRouter-specific usage details remain under `providerMetadata.openrouter.usage`.

## Local checks

Baseline on upstream `main` before migration:

```text
pnpm typecheck && pnpm build && pnpm test
Test Files 21 passed (21)
Tests 427 passed (427)
```

Final checks on this branch:

```text
pnpm stylecheck
Checked 133 files in 140ms. No fixes applied.

pnpm typecheck
tsc --noEmit

pnpm build
ESM dist/index.js     179.58 KB
ESM dist/internal/index.js     170.70 KB
DTS dist/index.d.ts 31.30 KB
DTS dist/internal/index.d.ts 25.47 KB

pnpm test:node
Test Files 21 passed (21)
Tests 427 passed (427)

pnpm test:edge
Test Files 21 passed (21)
Tests 427 passed (427)
```

`pnpm test:e2e` and `pnpm test:issues` were not run because they hit the real OpenRouter API and require `OPENROUTER_API_KEY`.

## Release notes

- Added a major changeset.
- Updated README and CHANGELOG for AI SDK v7, Node 22, ESM-only packaging, and v7 usage token details.

### CODE REVIEW

**Comment by Shreyaan** [comment]
Review/context after the final two follow-up commits:

- `1e01a81` updates CI and publish workflows from Node 20.x to 22.x, matching the new `engines.node >=22` requirement from AI SDK v7.
- `a3e097f` updates live e2e issue tests for AI SDK v7 semantics: `allowSystemInMessages: true` where those regression tests intentionally place cache-control system messages in `messages[]`, and `include: { responseBody: true }` where the test asserts `result.response.body`.
- Retired Anthropic model slugs were intentionally not changed in this PR; tracked separately in #512.
- Live-model edge cases left open for maintainers/follow-up: issue-194 (`grok` invalid JSON), issue-248 (`gemini` web-search empty), issue-438 (`gemini` reasoning redacted). These look model/API-behavior dependent rather than core v7 migration blockers.

Fresh local checks run on the final 3-commit branch:

- `pnpm stylecheck` ✅ `Checked 133 files in 154ms. No fixes applied.`
- `pnpm typecheck` ✅ `tsc --noEmit`
- `pnpm build` ✅ ESM build + DTS build succeeded
- `pnpm test:node` ✅ `Test Files 21 passed (21)`, `Tests 427 passed (427)`
- `pnpm test:edge` ✅ `Test Files 21 passed (21)`, `Tests 427 passed (427)`

As expected for a fork PR, GitHub Actions have not started yet because the first workflow run needs maintainer approval.

**Comment by Shreyaan** [comment]
Local CodeRabbit CLI review result:

Initial run found 3 issues, all addressed in follow-up commit `1486aef`:

- cross-platform build cleanup: changed `clean` away from POSIX-only `rm` now that `build` invokes `clean` directly
- V4 file URL validation: `{ type: "url" }` file data now accepts only `http:`, `https:`, and `data:` URLs; rejects `file:`/`blob:` before request construction
- completion streaming V4 contract: emits `{ type: "stream-start", warnings: [] }` before completion stream parts

Fresh local checks after the fix commit:

- `pnpm stylecheck` ✅ `Checked 133 files in 380ms. No fixes applied.`
- `pnpm typecheck` ✅ `tsc --noEmit`
- `pnpm build` ✅ ESM + DTS build succeeded
- `pnpm test:node` ✅ `Test Files 21 passed (21)`, `Tests 429 passed (429)`
- `pnpm test:edge` ✅ `Test Files 21 passed (21)`, `Tests 429 passed (429)`

Reran local CodeRabbit CLI against the full branch (`coderabbit review --agent --type all --base upstream/main`) after the fixes: ✅ `review_completed`, `findings: 0`.

File: package.json [hunk 1]
```diff
@@ -65,11 +64,11 @@
     "vitest": "3.2.4"
   },
   "peerDependencies": {
-    "ai": "^6.0.0",
-    "zod": "^3.25.0 || ^4.0.0"
+    "ai": "^7.0.0",
+    "zod": "^3.25.76 || ^4.1.8"
```
[the remark this answers is not in the document]
Comment (Shreyaan) [reply package.json:68]: Fixed in `2fe3abf`.

Verified the failure first:

```text
pnpm install --frozen-lockfile
ERR_PNPM_OUTDATED_LOCKFILE ... zod (lockfile: ^3.25.0 || ^4.0.0, manifest: ^3.25.76 || ^4.1.8)
```

Regenerated the lockfile and kept the diff to the single root importer specifier line:

```diff
- specifier: ^3.25.0 || ^4.0.0
+ specifier: ^3.25.76 || ^4.1.8
```

Fresh verification after the fix:

- `pnpm install --frozen-lockfile` ✅ `Lockfile is up to date, resolution step is skipped`
- `pnpm stylecheck` ✅ `Checked 133 files in 336ms. No fixes applied.`
- `pnpm typecheck && pnpm build && pnpm test:node && pnpm test:edge` ✅
  - `test:node`: `Test Files 21 passed (21)`, `Tests 429 passed (429)`
  - `test:edge`: `Test Files 21 passed (21)`, `Tests 429 passed (429)`

**Comment by Shreyaan** [comment]
Final local + live retest update:

Final PR-state checks run after reverting all temporary live-test slug changes:

- `pnpm format` ✅ `Formatted 133 files in 74ms. No fixes applied.`
- `pnpm install --frozen-lockfile` ✅ `Lockfile is up to date, resolution step is skipped`
- `pnpm stylecheck` ✅ `Checked 133 files in 103ms. No fixes applied.`
- `pnpm typecheck` ✅ `tsc --noEmit`
- `pnpm build` ✅ ESM + DTS build succeeded
- `pnpm test:node` ✅ `Test Files 21 passed (21)`, `Tests 429 passed (429)`
- `pnpm test:edge` ✅ `Test Files 21 passed (21)`, `Tests 429 passed (429)`

Live API checks with temporary local slug/output-cap substitutions:

- `pnpm test:e2e` ✅ `Test Files 12 passed (12)`, `Tests 26 passed (26)`
- Initial `pnpm test:issues` progressed to `38 passed` files, `145 passed` tests, `4 skipped`; failures were limited to 6 files.
- Targeted rerun of failed issue files with current/cheaper slugs and explicit output caps:
  - issue-194 (`x-ai/grok-4-fast` → `x-ai/grok-4.3`) ✅ 3/3 passed
  - issue-196 (`allowSystemInMessages`, current Anthropic slug, output cap) ✅ 1/1 passed
  - issue-248 (current/capped Anthropic web-search cases) ✅ 4 passed, 3 skipped
  - issue-453 (capped Anthropic signature cases) ✅ 4/4 passed
  - issue-190 still timed out: 2/2 timed out at 120s even with `openai/gpt-4.1-nano` and `maxOutputTokens: 100`
  - issue-438 still failed: Gemini response did not include `reasoning.encrypted` details, so both assertions expecting encrypted details failed

The temporary live-test substitutions were not committed; the branch is clean and still only contains the 5 PR commits.

**Comment by Shreyaan** [comment]
Follow-up for Perry's non-blocking stream-start symmetry note:

- Chat `doStream` now emits the same initial `{ type: "stream-start", warnings: [] }` part as completion `doStream`.
- The returned `warnings` array is shared with the stream-start part, matching the V4 stream warning surface.
- Updated exact chat stream expectations for text, tool, file, error, and parse-error paths.

Fresh verification after commit `b3e696b`:

- `pnpm format` ✅ `Formatted 133 files in 176ms. No fixes applied.`
- `pnpm install --frozen-lockfile` ✅ `Lockfile is up to date, resolution step is skipped`
- `pnpm stylecheck` ✅ `Checked 133 files in 120ms. No fixes applied.`
- `pnpm typecheck` ✅ `tsc --noEmit`
- `pnpm build` ✅ ESM + DTS build succeeded
- `pnpm test:node` ✅ `Test Files 21 passed (21)`, `Tests 429 passed (429)`
- `pnpm test:edge` ✅ `Test Files 21 passed (21)`, `Tests 429 passed (429)`

**Comment by Shreyaan** [comment]
Heads up: I force-pushed this branch only to fix commit attribution.

All commits are now authored by the personal `Shreyaan` account instead of my work account. The source tree is unchanged from the previous PR tip; this was a history/metadata rewrite only.

No new code changes were introduced by the force-push.

**Comment by yonatan-shorani** [comment]
Hi, any update on this PR?

**Comment by Shreyaan** [comment]
> Hi, any update on this PR?

waiting on any feedback or update 😅

**Comment by bdezso** [comment]
any update? 🔔

**Comment by ahmed-mkacher** [comment]
Just here to spread word that we also need this, this is a desired PR. thank you for your work! 🙌

**Comment by garshythoel** [comment]
Tagging @robert-j-y @Robinnnnn - apologies if not the right people, just saw y'all have committed recently

**Review by Robinnnnn — APPROVED** [review]

**Comment by Shreyaan** [comment]
Small follow-up pushed: while rerunning the live issue tests, I found one more AI SDK v7 semantics case in `issue-196`.

That test intentionally sends a `system` role inside `messages[]` to verify cache-control behavior. AI SDK v7 rejects system messages there unless `allowSystemInMessages: true` is set, matching the earlier fixes for the other cache-control issue tests. I added that flag only for this regression test.

Reverified locally after the push:

- `pnpm install --frozen-lockfile`
- `pnpm format`
- `pnpm stylecheck`
- `pnpm typecheck`
- `pnpm build`
- `pnpm test:node` — 432/432
- `pnpm test:edge` — 432/432

Could a maintainer approve the workflow run for the latest fork push when convenient?

**Comment by garshythoel** [comment]
Hey folks - still blocked here for an upgrade, Could a mantainer look ?

**Comment by DavertMik** [comment]
Please please please! :pray:

**Comment by Shreyaan** [comment]
Thanks everyone for the support here 🙏
@Robinnnnn @robert-j-y, could one of you please approve the workflow when convenient? Happy to address anything else needed to get this merged.

**Comment by robert-j-y** [comment]
This is merged, lets keep an eye out for any issues that may come up from the community!
